builder/builder.go

```diff
@@ -42,14 +49,78 @@ func New(privKey *ecdsa.PrivateKey, ownAddr *felt.Felt, bc *blockchain.Blockchai
 		bc:       bc,
 		vm:       builderVM,
 		newHeads: feed.New[*core.Header](),
+		log:      log,
 	}
 }
 
 func (b *Builder) Run(ctx context.Context) error {
-	<-ctx.Done()
+	if err := b.initPendingBlock(); err != nil {
+		return err
+	}
+	defer b.clearPending()
+
+	const blockTime = 5 * time.Second
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(blockTime):
+			if err := b.Finalise(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (b *Builder) initPendingBlock() error {
+	if b.pendingBlock.Block != nil {
+		return nil
+	}
+
+	bcPending, err := b.bc.Pending()
+	if err != nil {
+		return err
+	}
+
+	b.pendingBlock, err = utils.Clone[blockchain.Pending](bcPending)
```

Comment:
Is a round-trip through cbor necessary?

Reply:
This is the only generic enough way we have to deep copy a struct.

Reply:
It is a easy way of deep copying a struct. We can replace it with a dependency later on.